File: RemixedDungeon/src/main/java/com/watabou/pixeldungeon/actors/mobs/Goo.java

package com.watabou.pixeldungeon.actors.mobs;

import com.nyrds.Packable;
import com.nyrds.pixeldungeon.items.Treasury;
import com.nyrds.pixeldungeon.mechanics.NamedEntityKind;
import com.nyrds.pixeldungeon.ml.R;
import com.nyrds.platform.util.StringsManager;
import com.watabou.noosa.Camera;
import com.watabou.pixeldungeon.Badges;
import com.watabou.pixeldungeon.actors.Char;
import com.watabou.pixeldungeon.actors.CharUtils;
import com.watabou.pixeldungeon.actors.blobs.ToxicGas;
import com.watabou.pixeldungeon.actors.buffs.Buff;
import com.watabou.pixeldungeon.actors.buffs.Ooze;
import com.watabou.pixeldungeon.items.keys.SkeletonKey;
import com.watabou.pixeldungeon.sprites.CharSprite;
import com.watabou.pixeldungeon.sprites.GooSprite;
import com.watabou.pixeldungeon.utils.GLog;
import com.watabou.utils.Random;

import org.jetbrains.annotations.NotNull;

public class Goo extends Boss {

	private static final float PUMP_UP_DELAY	= 2.2f;
	
	public Goo() {
		hp(ht(68));
		exp = 9;
		baseDefenseSkill = 12;
		baseAttackSkill = 11;
		dr = 2;
		spriteClass = GooSprite.class;

		loot(Treasury.Category.POTION, 0.8f);
		
		addResistance( ToxicGas.class );

		collect(new SkeletonKey());
	}
	

	@Packable
	private boolean pumpedUp = false;

	@Override
	public int damageRoll() {
		if (pumpedUp) {
			dmgMin = 7;
			dmgMax = 21;
			return super.damageRoll();
		} else {
			dmgMin = 4;
			dmgMax = 11;
			return super.damageRoll();
		}
	}
	
	@Override
	public int attackSkill( Char target ) {
		return pumpedUp ? 26 : 11;
	}

	@Override
	public boolean act() {
		
		if (level().water[getPos()] && hp() < ht()) {
			heal(1,this);
		}
		
		return super.act();
	}
	
	@Override
    public boolean canAttack(@NotNull Char enemy) {
		return pumpedUp ? distance( enemy ) <= 2 : super.canAttack(enemy);
	}
	
	@Override
	public int attackProc(@NotNull Char enemy, int damage ) {
		if (Random.Int( 3 ) == 0) {
			Buff.affect( enemy, Ooze.class );
			enemy.getSprite().burst( 0x000000, 5 );
		}
		
		if (pumpedUp) {
			Camera.main.shake( 3, 0.2f );
		}
		
		return damage;
	}
	
	@Override
	public void doAttack(Char enemy) {
		if (pumpedUp || Random.Int( 3 ) > 0) {
			super.doAttack( enemy );
		} else {
			
			pumpedUp = true;
			spend( PUMP_UP_DELAY );
			
			((GooSprite)getSprite()).pumpUp();
			
			if (CharUtils.isVisible(this)) {
                getSprite().showStatus( CharSprite.NEGATIVE, StringsManager.getVar(R.string.Goo_StaInfo1));
                GLog.n(StringsManager.getVar(R.string.Goo_Info1));
			}
		}
	}
	
	@Override
	public boolean attack(@NotNull Char enemy ) {
		boolean result = super.attack( enemy );
		pumpedUp = false;
		return result;
	}
	
	@Override
	public boolean getCloser(int target) {
		pumpedUp = false;
		return super.getCloser( target );
	}

	@Override
	public void die(@NotNull NamedEntityKind cause) {
		super.die( cause );

		Badges.validateBossSlain(Badges.Badge.BOSS_SLAIN_1);

        yell(StringsManager.getVar(R.string.Goo_Info2));
	}
	
	@Override
	public void notice() {
		super.notice();
        yell(StringsManager.getVar(R.string.Goo_Info3));
	}

	@Override
	public boolean zap(@NotNull Char enemy) {
		pumpedUp = false;
		return true;
	}
}

File: scripts/mobs/Goo.lua
--[[
  Goo - Sewer boss (batch 17c-1, was actors/mobs/Goo.java).
  Boss die-flow (banner/music/unseal/key) rides the isBoss json key.
  Soaks in water to heal, pumps up (1/3 of adjacent turns) for one
  heavy strike with reach two, leaves ooze on hit. The pumped strike
  at range two is a blank zap that un-pumps - faithful to the java
  flow. pumpedUp rides the script data (serpent round-trips the save).
]]
local RPD = require "scripts/lib/commonClasses"
local mob = require "scripts/lib/mob"

local PUMP_UP_DELAY = 2.2
local function pumped(self)
    return mob.restoreData(self).pumpedUp
end

local function setPumped(self, value)
    mob.restoreData(self).pumpedUp = value
end

return mob.init{
    damageRoll = function(self)
        if pumped(self) then
            return math.random(7, 21)
        end
        return math.random(4, 11)
    end,

    attackSkill = function(self, target)
        if pumped(self) then
            return 26
        end
        return 11
    end,

    act = function(self)
        -- nil return: the java act runs as usual
        if self:hp() < self:ht() and RPD.Dungeon.level.water[self:getPos() + 1] then
            self:heal(1, self)
        end
    end,

    canAttack = function(self, enemy)
        if pumped(self) then
            return self:distance(enemy) <= 2
        end
        return nil
    end,

    attackProc = function(self, enemy, dmg)
        if enemy ~= nil then
            if math.random(3) == 1 then
                RPD.Buffs.Buff:affect(enemy, "Ooze")
                enemy:getSprite():burst(0x000000, 5)
            end
            if pumped(self) then
                RPD.shakeCamera(3, 0.2)
            end
        end
        return dmg
    end,

    doAttack = function(self, enemy)
        if enemy == nil or pumped(self) or math.random(3) > 1 then
            return false
        end

        setPumped(self, true)
        self:spend(PUMP_UP_DELAY)
        self:getSprite():playExtra("pump")

        if RPD.CharUtils:isVisible(self) then
            self:getSprite():showStatus(RPD.CharSprite.NEGATIVE, RPD.textById("Goo_StaInfo1"))
            RPD.GLog:n(RPD.textById("Goo_Info1"), {})
        end
        return true
    end,

    getCloser = function(self, target, ignorePets)
        setPumped(self, false)
        return false
    end,

    zap = function(self, enemy)
        -- pumped reach-two strike is a blank zap that un-pumps
        setPumped(self, false)
        return true
    end,

    die = function(self, cause)
        RPD.CharUtils:validateBossSlain("BOSS_SLAIN_1")
        self:yell(RPD.textById("Goo_Info2"))
    end,

    notice = function(self)
        self:yell(RPD.textById("Goo_Info3"))
    end,
}
